File: 2022/code/python/1.py
# load libraries
import pandas as pd
import numpy as np

# read data
df = pd.read_csv(
  './2022/data/1.txt', 
  names = ['value'],
  dtype = 'Int64',
  skip_blank_lines = False
)

# # count total number of candies for each elf
df = df\
.assign(elf = np.where(df["value"].isnull(), 1, 0).cumsum())\
.groupby("elf", as_index = False)\
.agg(value = ("value", sum))

# get maximum number of candies
df["value"].max()

# get sum of candies of the top 3 elves
df\
.nlargest(n = 3, columns = "value")\
.agg(value = ('value', sum))\
.iloc[0,0]

File: 2022/code/r/1.R
# load libraries
library(readr)
library(dplyr)

# read data
data <- read_delim(
  file = './2022/data/1.txt',
  delim = '\n',
  col_names = 'value',
  col_types = 'i',
  skip_empty_rows = FALSE
)

# count total number of candies for each elf
data <- data |> 
  mutate(elf = cumsum(is.na(value))) |> 
  count(elf, wt = value) 

# get maximum number of candies
data |> 
  summarise(n = max(n)) |> 
  pull()

# get sum of candies of the top 3 elves
data |> 
  slice_max(n = 3, order_by = n, with_ties = FALSE) |> 
  summarise(n = sum(n)) |> 
  pull()
